### PR TITLE
Pass resource version to storage List operation.

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref: #15945 

This is change in internal interfaces - the main implementation will come once this interface change is merged.

cc @smarterclayton

### CODE REVIEW

File: pkg/api/rest/rest.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ type Lister interface {
 	// This object must be a pointer type for use with Codec.DecodeInto([]byte, runtime.Object)
 	NewList() runtime.Object
 	// List selects resources in the storage which match to the selector.
-	List(ctx api.Context, label labels.Selector, field fields.Selector) (runtime.Object, error)
+	List(ctx api.Context, label labels.Selector, field fields.Selector, options *api.ListOptions) (runtime.Object, error)
```
Comment (smarterclayton) [inline pkg/api/rest/rest.go]: Typically this is untyped (GetterWithOptions accepts runtime.Object) and we support both the old and the new interface.  However, since List() is effectively standard and we don't have a use case for api.ListOptions varying between callers, I'm ok with making this change as is.

Can you add a TODO on this interface to remove label and field selectors (since they are part of api.ListOptions)?

File: pkg/registry/generic/etcd/etcd.go [hunk 2]
```diff
@@ -183,7 +183,15 @@ func (e *Etcd) ListPredicate(ctx api.Context, m generic.Matcher) (runtime.Object
 	}
 
 	trace.Step("About to list directory")
-	err := e.Storage.List(ctx, e.KeyRootFunc(ctx), filterFunc, list)
+	if options == nil {
+		// FIXME: Should we pass a different rv by default?
+		options = api.NewListOptions("0")
```
Comment (smarterclayton) [inline pkg/registry/generic/etcd/etcd.go]: This is probably ok for now.

File: pkg/api/rest/rest.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/api/rest/rest.go]: Sure - I want to remove it from both List and Watch (both here in and in the client).
I will add TODO tomorrow.

Comment (wojtek-t) [inline pkg/api/rest/rest.go]: @smarterclayton @lavalamp 
BTW - the reason I decided for that API was to make it consistent with DeleteOptions in GracefulDeleter.

Comment (smarterclayton) [inline pkg/api/rest/rest.go]: That's fine.  Delete and List are more similar than Gets to different subresources will be (proxy, logs, status are all different Get schemas)

File: pkg/api/helpers.go [hunk 3]
```diff
@@ -118,6 +118,12 @@ func NewDeleteOptions(grace int64) *DeleteOptions {
 	return &DeleteOptions{GracePeriodSeconds: &grace}
 }
 
+// NewListOptions returns a ListOptions indicating the resource should be
+// listed with at least given 'resourceVersion'.
+func NewListOptions(resourceVersion string) *ListOptions {
```
Comment (lavalamp) [inline pkg/api/helpers.go]: I never thought I'd argue against a constructor, but is a constructor for list options really useful? I'd expect it's better to just have people set the fields they care about. That way if you add a field you don't need to change the constructor and all invocations.

(caveat: the above doesn't apply if there's more to be done than just setting fields)

File: pkg/api/rest/rest.go (hunk 1 above)
Comment (lavalamp) [inline pkg/api/rest/rest.go]: I'd argue for ListOptions going to unversioned, like we chatted about earlier.

Also can label & field be moved into listOptions?

File: pkg/api/rest/rest.go [hunk 4]
```diff
@@ -62,7 +62,7 @@ type Lister interface {
 	// This object must be a pointer type for use with Codec.DecodeInto([]byte, runtime.Object)
 	NewList() runtime.Object
 	// List selects resources in the storage which match to the selector.
```
Comment (lavalamp) [inline pkg/api/rest/rest.go]: add that options may be nil

File: pkg/api/rest/rest.go (hunk 1 above)
Comment (lavalamp) [inline pkg/api/rest/rest.go]: I guess that's what Clayton said already. :)

Comment (wojtek-t) [inline pkg/api/rest/rest.go]: Yes - definitely I was to move label and field into listOptions (here and in client, etc.), but I would like to do it in a separate PR.

Comment (smarterclayton) [inline pkg/api/rest/rest.go]: Separate PR is fine for me, it's mostly busywork refactor.

We have the constructor for delete options because it's impossible to do
"&0" (take address of constant) in Go.  Case is weaker for other value
types.

File: pkg/api/helpers.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/api/helpers.go]: Yes - initially I though it will be, but it doesn't seem to be that useful.
Removing it.

File: pkg/api/rest/rest.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/api/rest/rest.go]: Done.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL

File: pkg/api/rest/rest.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/api/rest/rest.go]: OK - thanks.

@lavalamp - moving ListOptions to unversioned is orthogonal change - I will change that in a separate PR.

**Comment by wojtek-t** [comment]
@lavalamp - PTAL (moving to the top of your inbox :))

File: pkg/api/rest/rest.go [hunk 5]
```diff
@@ -61,8 +61,9 @@ type Lister interface {
 	// NewList returns an empty object that can be used with the List call.
 	// This object must be a pointer type for use with Codec.DecodeInto([]byte, runtime.Object)
 	NewList() runtime.Object
-	// List selects resources in the storage which match to the selector.
-	List(ctx api.Context, label labels.Selector, field fields.Selector) (runtime.Object, error)
+	// List selects resources in the storage which match to the selector. 'options' can be nil.
+	// TODO: Move 'label' and 'field' to 'options'.
+	List(ctx api.Context, label labels.Selector, field fields.Selector, options *api.ListOptions) (runtime.Object, error)
```
Comment (lavalamp) [inline pkg/api/rest/rest.go]: I think we'll want to move to unversioned.ListOptions eventually, but this is fine now.

**Comment by lavalamp** [comment]
LGTM
